libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp

```diff
@@ -1274,63 +1670,620 @@ void PeerConnectionImpl::OnLocalDescCreated(
+                                return tr->HasReceiver(receiver);
+                              });
+    if (it_tr != transceivers_.end()) {
+      RTC_DCHECK(media_kind == (*it_tr)->GetMediaKind());
+      RefPtr<Transceiver> transceiver = static_cast<Transceiver*>(it_tr->get());
+      return transceiver;
+    }
+  }
+
+  if (IsUnifiedPlan()) {
+    // The new remote track should already have a low-level implementation RTP
+    // transceiver from applying the remote description. But the wrapper for it
+    // was not created yet. Find the RTP transceiver of the RTP receiver,
+    // bearing in mind its mline index is not necessarily contiguous in the
+    // wrapper array.
+    auto transceivers = peer_->GetTransceivers();
+    auto it_impl = std::find_if(
+        transceivers.begin(), transceivers.end(),
+        [receiver](auto&& tr) { return tr->receiver() == receiver; });
+    if (it_impl == transceivers.end()) {
+      return Error(
+          Result::kNotFound,
+          "Failed to match RTP receiver with an existing RTP transceiver.");
+    }
+    rtc::scoped_refptr<webrtc::RtpTransceiverInterface> impl = *it_impl;
+    const int mline_index = (int)std::distance(transceivers.begin(), it_impl);
+
+    std::string name = ExtractTransceiverNameFromReceiver(receiver);
+
+    // Create a new transceiver wrapper for it
+    return CreateTransceiverUnifiedPlan(media_kind, mline_index,
+                                        std::move(name), std::move(impl));
+  } else {
+    RTC_DCHECK(IsPlanB());
+    // In Plan B, since there is no guarantee about the order of tracks, they
+    // are matched by stream ID (msid).
+    int mline_index = -1;
+    std::string name;
+    if (!ExtractMlineIndexAndNameFromReceiverPlanB(receiver, mline_index,
+                                                   name)) {
+      return Error(
+          Result::kUnknownError,
+          "Failed to associate RTP receiver with Plan B emulated mline index "
+          "for track pairing.");
+    }
+
+    // Create a new audio transceiver wrapper for it
+    return CreateTransceiverPlanB(media_kind, mline_index, std::move(name),
+                                  receiver);
+  }
+}
+
+ErrorOr<RefPtr<Transceiver>> PeerConnectionImpl::GetOrCreateTransceiver(
+    int mline_index,
+    webrtc::RtpTransceiverInterface* rtp_transceiver,
+    std::string name) {
+  // Convert and validate media kind
+  mrsMediaKind media_kind;
+  switch (rtp_transceiver->media_type()) {
+    case cricket::MediaType::MEDIA_TYPE_AUDIO:
+      media_kind = mrsMediaKind::kAudio;
+      break;
+    case cricket::MediaType::MEDIA_TYPE_VIDEO:
+      media_kind = mrsMediaKind::kVideo;
+      break;
+    default:
+      return Error(Result::kInvalidMediaKind, "Unknown SDP semantic");
+  }
+
+  // Find an existing transceiver wrapper which would have been created just
+  // a moment ago by the remote track added callback.
+  rtc::CritScope lock(&tracks_mutex_);
+  for (auto&& tr : transceivers_) {
```

Comment:
Do you need to iterate if you have the mline_index?

Reply:
No you don't, will fix.

Reply:
Fixed.